### PR TITLE
feat: add mint functionality for LSP

### PR DESCRIPTION
**Pull request by Tulun**
**Motivation**
Mint functionality works on contract.

**Summary**

Note: In order to get this working, you must spin up a fork of mainnet and set up the hardhat fork in metamask.

**Details**

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [x]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

**Review by daywiss — COMMENTED** [review]
have a couple small comments and nits, but the overall approach looks good. mainly:
1. i dont think the form input between changing long/short vs collateral amount works right in all cases
2. total minted value is not an accurate representation of your long/short token balance
3. we should not be conflating the display of bignumbers vs the math with big numbers. converting back and forth between big nums may introduce unpredictable results

File: components/lsp-form/MintForm.tsx
Comment (daywiss) [reply components/lsp-form/MintForm.tsx:60]: https://github.com/UMAprotocol/claim-dapp/blob/master/src/hooks/useRedeem.ts#L31

File: components/lsp-form/LSPForm.tsx
Comment (Tulun) [reply components/lsp-form/LSPForm.tsx]: good call!

File: components/lsp-form/LSPForm.tsx
Comment (Tulun) [reply components/lsp-form/LSPForm.tsx]: @mrice32 , I guess this would be the case, as ERC20 tokens can be any decs?

File: components/lsp-form/MintForm.tsx
Comment (Tulun) [reply components/lsp-form/MintForm.tsx:60]: Yeah I'm wondering if we should incorporate a different flow for approvals here. but I can default to infinite and do a quick check if need be.

File: components/lsp-form/MintForm.tsx
Comment (Tulun) [reply components/lsp-form/MintForm.tsx]: yeah it's a bit odd here, as the ratio will likely be a small integer.

Do we have any good utility functions for this math? Unfortunately, ethers BN doesn't allow floats.

File: components/lsp-form/LSPForm.tsx
Comment (mrice32) [reply components/lsp-form/LSPForm.tsx]: Yeah, the token decimals depends on the collateral decimals, so you definitely need to consider them when converting to a user-readable output.

File: components/lsp-form/MintForm.tsx
Comment (mrice32) [reply components/lsp-form/MintForm.tsx]: @Tulun we also need to be super careful with rounding. The contract always rounds the amount of collateral needed _up_.

File: components/lsp-form/Collateral.tsx
Comment (mrice32) [inline components/lsp-form/Collateral.tsx:28]: comment?

File: components/lsp-form/MintForm.tsx
Comment (mrice32) [reply components/lsp-form/MintForm.tsx:60]: @Tulun The way we did this in EMP tools was to approve for UINT_MAX (2^256 - 1). Only get the user to approve if their current approval is < UINT_MAX / 2.

File: components/lsp-form/useTokensCreatedEvents.ts
Comment (mrice32) [inline components/lsp-form/useTokensCreatedEvents.ts:28]: Comment?

File: components/lsp-form/LongShort.tsx
Comment (mrice32) [reply components/lsp-form/LongShort.tsx:57]: Agreed. Can we have some function that we call that's able to detect which box is the "selected" one so we avoid updating that one? Otherwise, there may be odd situations due to rounding, etc.

**Review by mrice32 — COMMENTED** [review]
Looks great!! Only minor comments to augment @daywiss's very detailed review.

Also, would be great if you could format the title of the PR as we do elsewhere: `feat: add mint functionality for LSP`!

File: components/lsp-form/LSPForm.tsx
Comment (Tulun) [reply components/lsp-form/LSPForm.tsx]: Sounds good, implementation of this added.

File: components/lsp-form/MintForm.tsx
Comment (Tulun) [reply components/lsp-form/MintForm.tsx]: Kept this as Big Number as follows @mrice32  @daywiss 

        // Need to send the correct amount based on the collateral pair ** the amount
        // User has specified in the input.
        // Get rid of float, then divide it out later.
        const ratio = ethers.BigNumber.from(1 / Number(collateralPerPair)).mul(
          ethers.BigNumber.from(TEN_POWER_18)
        );

        lspContract
          .create(weiAmount.mul(ratio).div(ethers.BigNumber.from(TEN_POWER_18)))

File: components/lsp-form/LSPForm.tsx
Comment (Tulun) [reply components/lsp-form/LSPForm.tsx]: Good call, I will keep all as BN and do math there.

File: components/text-input/TextInput.tsx
Comment (Tulun) [reply components/text-input/TextInput.tsx:24]: There is a default onChange handler inside, I could rename it though.

File: utils/convertFromWeiSafely.ts
Comment (Tulun) [reply utils/convertFromWeiSafely.ts:16]: This was pulled from the common lib but we can refactor!

File: components/lsp-form/LongShort.tsx
Comment (Tulun) [reply components/lsp-form/LongShort.tsx:57]: Good call.

File: components/lsp-form/useTokensCreatedEvents.ts
Comment (Tulun) [reply components/lsp-form/useTokensCreatedEvents.ts:28]: I am leaving this as a comment as to the shape of the events from the contract -- documentation. I can delete if you'd like though.

File: components/lsp-form/MintForm.tsx
Comment (Tulun) [reply components/lsp-form/MintForm.tsx]: Implementation like the claim-dapp is done. We could do UINT_MAX/2, although claim is just infinite.

File: components/lsp-form/MintForm.tsx
Comment (daywiss) [inline components/lsp-form/MintForm.tsx:111]: i still think thers a potential problem here, particularly with very small floats. why not allow a side effect on the setAmount input box, and have it setShort/setLong tokens directly?  so now you have a directional channel where when user is inputting in long/short inputs, that will setAmount and when user editing amount input, it sets long/short token.  you could also just combine long/short token amount into a single number since they should always be the same

File: components/lsp-form/LongShort.tsx
Comment (daywiss) [inline components/lsp-form/LongShort.tsx:68]: the label says balance, but you are showing tokens minted. just want to be sure what the intent here is, is it to show the users balance of long or short tokens? or is it to show how many have been minted through the interface? While for minting, i think we can assume a user will always mint equal amounts of long/short tokens. but their balance may be completely different, ie they may have more longs than shorts, since they can be traded on open market.

File: components/lsp-form/LongShort.tsx
Comment (daywiss) [inline components/lsp-form/LongShort.tsx:75]: again just want to be sure of what this is, because it might change the number here

File: components/lsp-form/MintForm.tsx
Comment (daywiss) [inline components/lsp-form/MintForm.tsx:90]: this might cause an unhandled error. can return promise and await or catch
```js
await lspContract
  .create(weiAmount.mul(ratio).div(ethers.BigNumber.from(10).pow(18)))
  .then((tx: any) => {
    return tx.wait(1)
  }).then(async ()=>{
    setAmount("");
    //...etc
  })

```

File: components/lsp-form/useTokensCreatedEvents.ts
Comment (mrice32) [reply components/lsp-form/useTokensCreatedEvents.ts:28]: Gotcha -- makes sense!

File: components/lsp-form/MintForm.tsx
Comment (mrice32) [reply components/lsp-form/MintForm.tsx]: I actually like grabbing the user's balance more. +1

File: utils/convertFromWeiSafely.ts
Comment (daywiss) [reply utils/convertFromWeiSafely.ts:16]: yea i dont think it will hurt to update it

File: components/lsp-form/MintForm.tsx
Comment (mrice32) [inline components/lsp-form/MintForm.tsx:83]: collateralPerPair is a string. It's unclear whether it's in wei or not, but if it is, this conversion can lose precision.

`1 / collateralPerPair` is also likely to lose precision in certain cases. We should do all of this math using a high-precision math library.

My suggestion is to decide up-front if you want all your numbers to be represented as true decimals using some sort of library like bignumber.js (which handles decimals) OR if you want to mirror solidity math and represent decimals as fixed point numbers that are multiplied up.

My suggestion for this line if you want to have these numbers use solidity-like math is the following

```ts
const toBN = ethers.BigNumber.from;
const ratio = toBN(10).pow(18).div(toBN(collateralPerPair));
```

Doing all multiplications first and all divisions last ensures that you lose as little precision as possible.

Note: to ensure you don't end up with rounding issues, you may have to compute the ratio in the create line below. Otherwise, the ratio may get rounded as an intermediate step and you may end up passing a slightly innacurate value to create. Something like the following may work a little better:

```ts
// Note: this assumes that collateralPerPairInWei is the raw value from the contract (i.e. 1e18 * decimal value)
const toBN = ethers.BigNumber.from;
const scaling = toBN(10).pow(18);
lspContract.create(weiAmount.mul(scaling).div(collateralPerPairInWei)).then(...)
```

**Review by mrice32 — COMMENTED** [review]
Looks good! A few comments. Most important one is around making sure we round the same way the contract does!

File: components/lsp-form/MintForm.tsx
Comment (Tulun) [reply components/lsp-form/MintForm.tsx:83]: CollateralPerPair is the value returned from the contract. I just converted it to a string. I can keep it as a big number if you'd prefer.

File: pages/testing.tsx
Comment (Tulun) [reply pages/testing.tsx:1]: It's for building the component independently of the view itself. We decided on this instead of having it started as integrated.

File: components/lsp-form/MintForm.tsx
Comment (Tulun) [reply components/lsp-form/MintForm.tsx:111]: Yeah totally reasonable, I can use the same flow as the other 2 and avoid useEffect altogether. It also self contains the update close to the actual input which is actually nice.

File: components/lsp-form/LongShort.tsx
Comment (Tulun) [reply components/lsp-form/LongShort.tsx]: Swapped for balances from the contracts.

File: components/lsp-form/LongShort.tsx
Comment (Tulun) [reply components/lsp-form/LongShort.tsx]: Makes sense, done.

File: components/lsp-form/LongShort.tsx
Comment (Tulun) [reply components/lsp-form/LongShort.tsx]: Updated to proper balance values of contracts.

File: components/lsp-form/LongShort.tsx
Comment (Tulun) [reply components/lsp-form/LongShort.tsx]: Same as above.

File: components/lsp-form/MintForm.tsx
Comment (Tulun) [reply components/lsp-form/MintForm.tsx]: Done!

File: components/lsp-form/MintForm.tsx
Comment (Tulun) [reply components/lsp-form/MintForm.tsx:42]: removed

File: pages/testing.tsx
Comment (daywiss) [inline pages/testing.tsx:77]: i feel like all these promises should be caught, but maybe its ok, not sure

**Review by daywiss — APPROVED** [review]
this is looking good, i've got a couple nits which might be worth addressing but going to approve under the assumption you will resolve all matts issues.

File: components/lsp-form/MintForm.tsx
Comment (mrice32) [reply components/lsp-form/MintForm.tsx:83]: It looks like you're converting it to a decimal before setting it, though. So it's the value returned by the contract divided by 1e18, right?

Either way, we need to be careful about precision loss (see above). This math needs to be updated to avoid precision loss or it will be susceptible to rounding errors that cause reverts.

File: pages/testing.tsx
Comment (mrice32) [reply pages/testing.tsx:1]: Oh I see. So it's just a placeholder. Got it.

**Review by mrice32 — COMMENTED** [review]
Looks good! I think the rounding errors discussed here: https://github.com/UMAprotocol/umaverse/pull/12#discussion_r674739847 just need to be fixed.

**Comment by Tulun** [comment]
> Looks good! I think the rounding errors discussed here: [#12 (comment)](https://github.com/UMAprotocol/umaverse/pull/12#discussion_r674739847) just need to be fixed.

@mrice32  fix implemented, give it a quick look over :)

File: components/lsp-form/MintForm.tsx
Comment (mrice32) [inline components/lsp-form/MintForm.tsx:67]: Is amount a perfect-precision decimal string? Is it possible there's any precision loss in amount before we get here?

File: components/lsp-form/MintForm.tsx
Comment (mrice32) [inline components/lsp-form/MintForm.tsx:88]: I think this will break if collateralPerPair is < 1e18 (aka less than one unit of collateral per pair of tokens). In that case, the ratio will round to 0 and you will mint 0 tokens no matter what the user inputs. What you want is to do _all_ the multiplication first and then _all_ the division at the end to minimize the rounding. Note: this will mean you will not be able to have well-reasoned intermediate values, like ratio, so the code will be a little less understandable :/.

```ts
// All operations that make the number larger come first. All the operations that make the number smaller come last.
const mintAmount = weiAmount.mul(scaledtoWei).div(collateralPerPair);
lspContract.create(mintAmount).then(...);
```

**Review by mrice32 — COMMENTED** [review]
Looks awesome! A few comments on rounding

**Review by mrice32 — APPROVED** [review]
LGTM!

**Pull request merged**
